src/System.Memory/src/System/Buffers/BuffersExtensions.cs

```diff
@@ -72,18 +72,11 @@ public static void Write<T>(this IBufferWriter<T> bufferWriter, ReadOnlySpan<T>
 
             while (source.Length > 0)
             {
-                int writeSize = destination.Length;
-
-                if (destination.Length == 0)
-                {
-                    writeSize = Math.Min(source.Length, bufferWriter.MaxBufferSize);
-                    destination = bufferWriter.GetSpan(writeSize);
-                }
-
+                int writeSize = Math.Min(destination.Length, source.Length);
                 source.Slice(0, writeSize).CopyTo(destination);
                 bufferWriter.Advance(writeSize);
                 source = source.Slice(writeSize);
-                destination = default;
+                destination = bufferWriter.GetSpan(source.Length);
```

Comment:
Can you please explain what this change fixes? What was wrong with the previous implementation?

Reply:
It fixes 
>    Aside from the first time through the loop, this will always be true (since destination = default). Consider special casing the first iteration outside the loop.

Without special casing anything.

Reply:
Yes, but now we are doing a lot of extra work if writeSize == 0 in the first iteration.

All these should be no-op when writeSize == 0 (which only happens if destination.Length == 0).
```C#
source.Slice(0, writeSize).CopyTo(destination);
bufferWriter.Advance(writeSize);
source = source.Slice(writeSize);
```

In terms of optimization, this approach from @benaadams is pretty good since there aren't any unnecessary operations when not needed:
https://github.com/dotnet/corefx/pull/27229#discussion_r170407410

Reply:
@benaadams Ugh, you are right.

Reply:
`writeSize` would not be `0` on the first iteration. `GetSpan` should always return non empty span.

Reply:
> `writeSize` would not be `0` on the first iteration. `GetSpan` should always return non empty span.

Ah yes, good point.

Reply:
@benaadams continue in do/while checks the exit condition.

Reply:
cc @AndyAyersMS

Reply:
@ahsonkhan what was it you thought I should look at?

Reply:
This method in BuffersExtensions:
https://github.com/dotnet/corefx/pull/27229/files#diff-321812293ad3b9aa1516f25eb68f8fbeR61
```C#
public static void Write<T>(this IBufferWriter<T> bufferWriter, ReadOnlySpan<T> source)
```

Is there a cleaner way to implement this that is just as optimal (or better)?

Comment thread for context: https://github.com/dotnet/corefx/pull/27229#discussion_r170371535

Reply:
> Is there a cleaner way to implement this that is just as optimal (or better)?

Yes, with a second `ref struct` that holds the span as state to avoid the virtual calls https://github.com/aspnet/Common/blob/46abbd1a136ae4a4bb93d17759ac4c0bbf59dee4/shared/Microsoft.Extensions.Buffers.Sources/BufferWriter.cs

Reply:
I don't think it would help with a single Write<T>, but if you have multiple of those BufferWriter is the way to go.

Reply:
Yeah, extension `Write` is for discoverability/convenience (else people will wonder how they write to it); then second struct is for perf. Will likely be other versions like an encoding writer (ala StreamWriter etc)?